Issue #2427: added customizable javadoc tokens

rnveach opened the pull request:
Issue #2427.

Added acceptable and required javadoc tokens. Code was mostly copied from `TreeWalker` which this class is mimicking.
All javadoc checks required `getRequiredJavadocTokens` to be added.

Questions:
1)
The issue made note of an [area that had documentation on this already but was removed](https://github.com/checkstyle/checkstyle/blob/fff24e89001ce2ad9c30d1f573639cd640183669/src/xdocs/config_javadoc.xml#L978).
Was that documentation valid that the 2 tokens are optional?
Are there any other optional tokens for the javadoc checks? I didn't see anything jump out when glancing over.

Some things to note:
1)
We needed some place to validate that the javadoc tokens set for the check are correct and require us to be able to throw a `CheckstyleException`. I did this in `init` and so had to add the exception to the throws clause, and this is why there is an update in `FallThroughCheck`. I don't see any other place we can do this.
I brought this up before where [none of the `AbstractCheck` methods allow throwing an exception](https://github.com/checkstyle/checkstyle/blob/25a37e5049b7816b34c552899841a978efc37a63/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java#L126-L169), even `CheckstyleException` and I don't think we should use runtime exceptions.
If your ok with it, I think we should change all the methods to be able to throw `CheckstyleException`s.

2)
Javadoc's `walk` method changed slightly because I think there is a bug with `waitsForProcessing` and when we loop `while (curNode != null && toVisit == null) {` more than once since the value isn't updated.
I will try to find a test to show this and add it.

3)
`AtClauseOrderCheck` changed slightly to not loop through the children, but to use the visit token's power. This would be helpful if the token appears in other places than a child of the root.

romani commented:
> Was that documentation valid where the 2 tokens are optional?

I doubt it was valid. It was imaginary config. 

> Are there any other optional tokens for the javadoc checks?

We do not have. We have only 6 Checks that extends AbstractJavadocCheck:
AtclauseOrderCheck
JavadocParagraphCheck
JavadocTagContinuationIndentationCheck
NonEmptyAtclauseDescriptionCheck
SingleLineJavadocCheck
SummaryJavadocCheck

All of them were created to cover Google style, right after ANTLR grammar was introduced. Almost all of them have are simple Checks. Only NonEmptyAtclauseDescriptionCheck could have some reasons to be customized. But it was missed from our attention during implementation as no customization was required for Google style.

romani commented:
> even CheckstyleException and I don't think we should use runtime exceptions.

we should not pollute method signatures with checked exceptions internally.
Checked exception give no extra details in comparison to not specifying it at all. It is already clear that there is possibility of exception in code. That is why non of Checks have it - when can throw all possible exceptions and TreeWalker convert them to CheckstyleException to deliver to outside of library as some exception that happened in checkstyle to easily let main code distinguish it and process.
I would use IllegalStateException , to let it fly to TreeWalker as any other exception in Checks.

> I will try to find a test to show this and add it.

Such unclear fixes should beeasily caught by cobertura. If you do not have problem, that is strange.

> AtClauseOrderCheck changed slightly to not loop through the children, but to use the visit token's power. This would be helpful if the token appears in other places than a child of the root.

 hmmm, it is always a decision point to make Check stateless or stateful (you added a field - it become stateful). I always guided other to make Check stateless is there is no performace hit by it in comparison to statefull implementation.
In the future:  Stateless Checks will be marked by us and TreeWalker will run them in MT mode, stateful Checks will run as they run right now.

rnveach commented:
> I doubt it was valid. It was imaginary config. 
> We do not have. 

Then I will leave validating javadoc tokens till later, as any code I write for it now will remain untested until that time.

I reverted changes to AtClauseOrderCheck.

>  Stateless Checks will be marked by us and TreeWalker will run them in MT mode, stateful Checks will run as they run right now.

I will start an email chain discussing this with you.

> we should not pollute method signatures with checked exceptions internally.

`CheckstyleException` is ours, so I don't really consider it polluting by adding it. My main reason for wanting it there is because I dislike wrapping exception upon exception multiple times making the final printed exception overly complex. But I am also unsure when something should and shouldn't be a runtime exception.
I reverted the changes and used `IllegalStateException`.

> Such unclear fixes should be easily caught by cobertura. If you do not have problem, that is strange.

Cobertura just checks for dead code. It doesn't check for logic errors, which this is. Just because a method runs and returns a result, doesn't mean it returns the correct result.
I added the check `testVisitLeaveToken`. If you comment out these 3 lines in `AbstractJavadocCheck`:
````
                 if (toVisit == null) {
                     curNode = curNode.getParent();
+                    if (curNode != null) {
+                        waitsForProcessing = shouldBeProcessed(curNode);
+                    }
````
you will get more calls to `leaveJavadocToken` than to `visitJavadocToken`. They should be 1 to 1. The reason for the issue is `waitsForProcessing` is a variable with a value based on `curNode`. When `curNode` changes, `waitsForProcessing` should too, but it wasn't.
Without these lines, the test will fail with: `AssertionError: expected:<406> but was:<464>`

romani commented:
> I dislike wrapping exception upon exception multiple times making the final printed exception overly complex with Caused by

yes, but unpleasant code is more often read by engineers.

=====================

Code looks good.

please update http://checkstyle.sourceforge.net/writingjavadocchecks.html#Customize_token_types_in_Javadoc_Checks and we good to merge.

rnveach commented:
@romani Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/NonEmptyAtclauseDescriptionCheck.java:
```diff
@@ -54,6 +54,11 @@
     @Override
+    public int[] getRequiredJavadocTokens() {
+        return getDefaultJavadocTokens();
+    }
```
hmm, the whole reason of the issue was to customize NonEmptyAtclauseDescription , but with such method we do not allow customization. 
required should be empty array.

rnveach replied:
> the whole reason of the issue was to customize NonEmptyAtclauseDescription 

I took your reply [here](https://github.com/checkstyle/checkstyle/pull/3512#issuecomment-255626830) to mean there were no optional tokens for any checks.

> > Was that documentation valid where the 2 tokens are optional?

> I doubt it was valid. It was imaginary config. 

> > Are there any other optional tokens for the javadoc checks?

> We do not have.

Done in next update.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTagContinuationIndentationCheck.java:
```diff
@@ -73,6 +73,11 @@ public void setOffset(int offset) {
 
     @Override
+    public int[] getRequiredJavadocTokens() {
+        return getDefaultJavadocTokens();
```
there should be getAcceptableJavadocTokens usage. We require all. 
Please fix in all other cases.

rnveach replied:
Done.

rnveach commented:
@romani Changes were addressed.
Allowing optional javadoc tokens in `NonEmptyAtclauseDescriptionCheck` required me to implement this code in `XDocsPagesTest` so they would appear in the xdocs.
